event-dispatcher/src/main/java/io/sk8s/event/dispatcher/EventDispatchingHandler.java

```diff
@@ -41,30 +51,122 @@
+
+	private final SpelExpressionParser spelExpressionParser = new SpelExpressionParser();
 
 	@EventListener
-	public void onFunctionRegistered(ResourceAddedEvent<XFunction> event) {
+	public void onFunctionRegistered(ResourceAddedOrModifiedEvent<XFunction> event) {
 		XFunction functionResource = event.getResource();
 		String functionName = functionResource.getMetadata().getName();
 		this.functions.put(functionName, functionResource);
-		this.launcher.dispatch(functionResource);
+		this.kafkaConsumerMonitor.beginTracking(functionName, functionResource.getSpec().getInput());
+
+		updateWakeUpInterval();
+		if (!running) {
+			running = true;
+			scalingThread.start();
+		}
+		synchronized (mutex) {
+			this.mutex.notify();
+		}
+
 		logger.info("function added: " + functionName);
 	}
 
+
 	@EventListener
 	public void onFunctionUnregistered(ResourceDeletedEvent<XFunction> event) {
 		XFunction functionResource = event.getResource();
 		String functionName = functionResource.getMetadata().getName();
-		this.kubernetesClient.extensions().jobs().withLabel("function", functionName).delete();
 		this.functions.remove(functionName);
+
+		this.kafkaConsumerMonitor.stopTracking(functionName, functionResource.getSpec().getInput());
+
+		updateWakeUpInterval();
+		synchronized (mutex) {
+			this.mutex.notify();
+		}
+
 		logger.info("function deleted: " + functionName);
 	}
+
+	@PreDestroy
+	public void tearDown() {
+		running = false;
+		synchronized (mutex) {
+			mutex.notify();
+		}
+		kafkaConsumerMonitor.close();
+	}
+
+	/**
+	 * Updates the wake up interval of the scaling thread to be the minimal desired value of all
+	 * functions, or zero (sleep forever) if there is no function.
+	 */
+	private void updateWakeUpInterval() {
+		this.scalingFrequencyMs = this.functions.values().stream()
+				.mapToInt(f -> f.getSpec().getScalingStrategy() == null ? 1_000
+						: f.getSpec().getScalingStrategy().getMaxUpdatePeriodMilliSeconds())
+				.min().orElse(0);
+	}
+
+	private class ScalingMonitor implements Runnable {
+
+		@Override
+		public void run() {
+			while (running) {
+
+				Map<KafkaConsumerMonitor.TopicGroup, List<KafkaConsumerMonitor.Offsets>> offsets = kafkaConsumerMonitor.compute();
+				logOffsets(offsets);
+				Map<String, Long> lags = offsets.entrySet().stream()
+						.collect(Collectors.toMap(
+							e -> e.getKey().group,
+							e -> e.getValue().stream()
+									.mapToLong(KafkaConsumerMonitor.Offsets::getLag)
+									.max().getAsLong()
+						));
+				functions.values().parallelStream().forEach(
+						f -> {
+							String expression = f.getSpec().getScalingStrategy() != null ? f.getSpec().getScalingStrategy().getReplicasExpression() : "1";
+							Expression spel = spelExpressionParser.parseExpression(expression);
+							EvaluationContext context = new StandardEvaluationContext();
+							context.setVariable("lag", lags.get(f.getMetadata().getName()));
+							int replicas = spel.getValue(context, Integer.class);
```

Comment:
I'm trying to see if I understand how this is working with the current example (using `#lag` as the expression)... that would create N replicas for N unconsumed events? Given that each instance can process multiple messages, that seems unnecessary in most cases. Also, there doesn't seem to be any relationship between the instance being created and the partition on which lag exists, correct?

Reply:
Yes this would. Obviously this is disproportionate, and we'd use some function (ah ah) to bring this number in a reasonable range. Including division, logarithm, Math.max, etc.

Reply:
> Also, there doesn't seem to be any relationship between the instance being created and the partition on which lag exists, correct

Correct. I didn't look into that yet (seemed orthogonal to pure lag detection / scaling).
But some rebalancing should occur automatically, shouldn't it?

Reply:
for simple load-balancing cases, yes... but we should (eventually) support more fine-grained control (if there are events on 2 partitions of a topic having 10 partitions, we would ideally start 2 consumers and handle partition assignment)... or another way to consider it with the current implementation is that starting 10 instances will not be any better than 2 instance in such a case

Reply:
regarding the various ways to determine the scaling value, I think initially we should just implement one approach and not worry about exposing it to users as a configurable option yet (that can come later)

Reply:
I agree. This is also why I was interested in the HPA definition which has `min, max, targetMetricValue`. Still need to look into it in more details to see if the scaler "understands" what it's doing or is just assuming more pods = reduced metric value.
Here we'd strive for target=0